docs/source/user-guide/guide_file_fetchers.rst

```diff
@@ -0,0 +1,61 @@
+- Implement ``FileFetcher\Processor\ProcessorInterface`` as a custom processor for ``FileFetcher``. Copy or subclass ``FileFetcher\Processor\Local`` and/or ``FileFetcher\Processor\Remote`` if it makes things easier.
+- Create a ``FileFetcherFactory`` class which instantiates a ``FileFetcher`` using configuration specifying the new processor in the processors array. Always merge in the configuration supplied to ``getInstance()``, as well.
+- Specify this new ``FileFetcherFactory`` as a service which decorates ``dkan.common.file_fetcher``.
+
+How to:
+=======
+
+To implement a new file processor, a custom file-fetcher processor class should be created, implementing ``FileFetcher\Processor\ProcessorInterface``. This class could subclass ``FileFetcher\Processor\Remote`` or ``FileFetcher\Processor\Local``, or be a new implementation. See the section below, "How processors work" for some more implementation details.
+
+A ``FileFetcherFactory`` class should then be created. The new factory should configure ``FileFetcher`` to use the custom processor. This is done by merging configuration for your new processor within the ``$config`` parameter to ``getInstance()``, something like this:
+
+    .. code-block:: php
+
+      public function getInstance(string $identifier, array $config = []) {
+        // Add OurProcessor as a custom processor.
+        $config['processors'] = array_merge(
+          [OurProcessor::class],
+          $config['processors'] ?? []
+        );
+        // Get the instance from the decorated factory, using our modified config.
+        return $this->decoratedFactory->getInstance($identifier, $config);
+      }
+
+It is also very important to declare your new factory class as a service. You accomplish this by decorating ``dkan.common.file_fetcher`` in your module's ``*.services.yml`` file, something like this:
+
+    .. code-block:: yaml
+
+      our_module.file_fetcher:
+        class: Drupal\our_module\FileFetcher\FileFetcherFactory
+        decorates: dkan.common.file_fetcher
+        arguments: ['@our_module.file_fetcher.inner']
+
+Now whenever DKAN uses the ``dkan.common.file_fetcher`` service, your file fetcher factory will be used instead.
+
+How processors work
+===================
+
+A little discourse on how file-fetcher decides which processor to use...
+
+File fetcher 'knows about' two processors by default: ``FileFetcher\Processor\Local`` and ``FileFetcher\Processor\Remote``. It also knows about whichever custom processor class names you configured in the ``processors`` array in configuration.
+
+When you ask a file fetcher object to perform the transfer (using the ``copy()`` method), it will instantiate all the different types of processors it knows about.
+
+Then it will loop through them and use the ``isServerCompatible()`` method to determine if the given ``source`` is suitable for use with that processor object. The file fetcher will use the first processor that answers ``TRUE``.
```

Comment:
One other thing - there is no `FileFetcher::copy()` method, though that's implied here, right? What there actually is is `FileFetcher::getProcessor()`, which iterates through the array and then runs `ProcessorInterfacecopy()` after finding a suitable processor class, right?

```suggestion
Processor negotiation
===================

The file-fetcher knows about two processors by default: ``FileFetcher\Processor\Local`` and ``FileFetcher\Processor\Remote``. It also knows about whichever custom processor class names you configured in the ``$processors`` array in your custom `FileFetcherFactory`.

When you ask a `FileFetcher` object to perform the transfer (using the ``copy()`` method), it will loop through `$processors` and use the ``ProcessorInterface::isServerCompatible()`` method from each to determine if the given ``source`` URL is suitable for use with that processor. The file-fetcher will use the first processor that answers `true`.
```